libs/testproxy/server.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/databricks/databricks-sdk-go/config"
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type ProxyServer struct {
+	*httptest.Server
+
+	t      testutil.TestingT
+	router *mux.Router
+
+	apiClient        *client.DatabricksClient
+	RequestCallback  func(request *testserver.Request)
+	ResponseCallback func(request *testserver.Request, response *testserver.EncodedResponse)
+}
+
+// This creates a reverse proxy server that sits in front of a real Databricks
+// workspace. This is useful for recording API requests and responses in
+// integration tests.
+//
+// Note: We cannot simply proxy the request from a localhost URL to a real
+// workspace. This is because auth resolution in the Databricks SDK relies
+// what the URL actually looks like to determine the auth method to use.
+// For example, in OAuth flows, the SDK can make requests to different Microsoft
+// OAuth endpoints based on the nature of the URL.
+// For reference, see:
+// https://github.com/databricks/databricks-sdk-go/blob/79e4b3a6e9b0b7dcb1af9ad4025deb447b01d933/common/environment/environments.go#L57
+func New(t testutil.TestingT) *ProxyServer {
+	router := mux.NewRouter()
+	server := httptest.NewServer(router)
+	t.Cleanup(server.Close)
+
+	s := &ProxyServer{
+		Server: server,
+		t:      t,
+		router: router,
+	}
+
+	// Create an API client using the current authentication context.
+	// In CI test environments this would read the appropriate environment
+	// variables.
+	var err error
+	s.apiClient, err = client.New(&config.Config{})
+	require.NoError(t, err)
+
+	// Set up the proxy handler as the default handler for all requests.
+	router.NotFoundHandler = http.HandlerFunc(s.proxyToCloud)
+	return s
+}
+
+func (s *ProxyServer) reqBody(r testserver.Request) any {
+	// The SDK expects the query parameters to be specified in the "request body"
+	// argument for GET, DELETE, and HEAD requests in the .Do method.
+	if r.Method == "GET" || r.Method == "DELETE" || r.Method == "HEAD" {
+		queryParams := make(map[string]any)
+		for k, v := range r.URL.Query() {
+			queryParams[k] = v[0]
+		}
+		return queryParams
+	}
+
+	// The SDK does not support directly passing a JSON serialized request
+	// body. So we convert it to a map if the body is a JSON object.
+	if json.Valid(r.Body) {
+		m := make(map[string]any)
+		err := json.Unmarshal(r.Body, &m)
+		assert.NoError(s.t, err)
+		return m
+	}
+
+	// Otherwise, return the raw body.
+	return r.Body
+}
+
+func (s *ProxyServer) proxyToCloud(w http.ResponseWriter, r *http.Request) {
+	request := testserver.NewRequest(s.t, r, nil)
+	if s.RequestCallback != nil {
+		s.RequestCallback(&request)
+	}
+
+	headers := make(map[string]string)
+	for k, v := range r.Header {
+		// Authorization headers will be set by the SDK. Do not pass them along here.
+		if k == "Authorization" {
+			continue
+		}
+		// The default HTTP client in Go sets the Accept-Encoding header to
```

Comment:
Why do we care about default HTTP client? We use SDK both for request to proxy and requests from proxy, right?

Reply:
The SDK uses the stdlib HTTP client. So the flow looks like:
test run ->  sdk1 -> proxy -> sdk2 -> workspace. 

"Accept-Encoding" is set by SDK1. I recall this having adverse effects.

Reply:
I see. What happens if you don't set, does it interferes with the default?